### PR TITLE
move some GetServices from constructor into CreateEditorInstance()

### LINKED ISSUE #27741: AbstractEditorFactory should be lazier in its MEF / service consumption
Originally filed by @ryanmolden as [DevDiv 487203](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/487203):

> This QS in the ctor of AbstractEditorFactory for SComponentModel and then asking it for several MEF types, is done synchronously during solution load. The service fetch and MEF import should be done much more lazily to move it outside the solution open path, or at least co-locate it with the actual creation of an editor, i.e. just in time.
> 
>    msenv.dll!CGlobalServiceProvider::QueryService(const _GUID & rsid, const _GUID & riid, void * * ppvObj) Line 1842 C++ Symbols loaded.
>   msenv.dll!CSPWrapper::QueryService(const _GUID & rsid, const _GUID & riid, void * * ppvObj) Line 2123 C++ Symbols loaded.
>   Microsoft.VisualStudio.Shell.15.0.ni.dll!059a555c() Unknown No symbols loaded.
>   [Frames below may be incorrect and/or missing, native debugger attempting to walk managed call stack]  Annotated Frame
>   [Managed to Native Transition]  Annotated Frame
>   Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.ServiceProvider.QueryService(System.Guid guid, System.Type serviceType, bool setShellErrorInfo, out object service) Line 230 C# Symbols loaded.
>   Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.ServiceProvider.GetService(System.Type serviceType, bool setShellErrorInfo) Line 100 C# Symbols loaded.
>   Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.Package.GetService(System.Type serviceType) Line 886 C# Symbols loaded.
>   Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.Package.System.IServiceProvider.GetService(System.Type serviceType) Line 1613 C# Symbols loaded.
> Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.AbstractEditorFactory.AbstractEditorFactory(Microsoft.VisualStudio.Shell.Package package) Line 39 C# Symbols loaded.
>   Microsoft.VisualStudio.LanguageServices.CSharp.dll!Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpPackage.CreateEditorFactories() Line 151 C# Symbols loaded.
>   Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage&lt;Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpPackage, Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpLanguageService&gt;.Initialize() Line 39 C# Symbols loaded.
>   Microsoft.VisualStudio.LanguageServices.CSharp.dll!Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpPackage.Initialize() Line 78 C# Symbols loaded.
>   Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.Package.Microsoft.VisualStudio.Shell.Interop.IVsPackage.SetSite(Microsoft.VisualStudio.OLE.Interop.IServiceProvider sp) Line 2151 C# Symbols loaded.
> &lt;snip&gt;

### PR DESCRIPTION
**Pull request by JieCarolHu**
Fixes #27741 

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes
https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/487203

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

**Comment by jinujoseph** [comment]
test windows_release_vs-integration_prtest

File: src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs [hunk 1]
```diff
@@ -29,18 +28,13 @@ internal abstract partial class AbstractEditorFactory : IVsEditorFactory, IVsEdi
...
         private Microsoft.VisualStudio.OLE.Interop.IServiceProvider _oleServiceProvider;
-        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
-        private readonly IContentTypeRegistryService _contentTypeRegistryService;
         private readonly IWaitIndicator _waitIndicator;
         private bool _encoding;
 
         protected AbstractEditorFactory(Package package)
         {
             _package = package ?? throw new ArgumentNullException(nameof(package));
             _componentModel = (IComponentModel)ServiceProvider.GetService(typeof(SComponentModel));
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs]: ❗️ The `IComponentModel` should be provided as an argument to this constructor rather than use `GetService` here.

:memo: #26787 will switch this from `GetService` to `GetServiceAsync`.

File: src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs [hunk 2]
```diff
@@ -102,7 +88,7 @@ int IVsEditorFactory.Close()
             // to properly handle multiple windows open for the same document.
             if (punkDocDataExisting != IntPtr.Zero)
             {
-                object docDataExisting = Marshal.GetObjectForIUnknown(punkDocDataExisting);
+                var docDataExisting = Marshal.GetObjectForIUnknown(punkDocDataExisting);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs]: :memo: Still seeing a bunch of unrelated formatting changes in PRs. Please submit them as a separate PR to unify the style before working, or disable the editor feature and omit them.

File: src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs [hunk 3]
```diff
@@ -29,18 +28,13 @@ internal abstract partial class AbstractEditorFactory : IVsEditorFactory, IVsEdi
...
         private bool _encoding;
 
         protected AbstractEditorFactory(Package package)
         {
             _package = package ?? throw new ArgumentNullException(nameof(package));
             _componentModel = (IComponentModel)ServiceProvider.GetService(typeof(SComponentModel));
-
-            _editorAdaptersFactoryService = _componentModel.GetService<IVsEditorAdaptersFactoryService>();
-            _contentTypeRegistryService = _componentModel.GetService<IContentTypeRegistryService>();
             _waitIndicator = _componentModel.GetService<IWaitIndicator>();
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs]: ❗️ This field should be lazily initialized as well per the original bug report (avoid unnecessary calls to `GetService` in this constructor).

File: src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs (hunk 2 above)
Comment (JieCarolHu) [reply src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs]: I am dogfooding code cleanup :)
 will use different PR next time

File: src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs [hunk 4]
```diff
@@ -29,19 +28,12 @@ internal abstract partial class AbstractEditorFactory : IVsEditorFactory, IVsEdi
...
         private Microsoft.VisualStudio.OLE.Interop.IServiceProvider _oleServiceProvider;
-        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
-        private readonly IContentTypeRegistryService _contentTypeRegistryService;
-        private readonly IWaitIndicator _waitIndicator;
         private bool _encoding;
 
-        protected AbstractEditorFactory(Package package)
+        protected AbstractEditorFactory(Package package, IComponentModel componentModel)
         {
             _package = package ?? throw new ArgumentNullException(nameof(package));
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs]: It looks like we can probably eliminate the package being passed entirely. The only use I'm seeing (assuming I'm not missing something) is we provide it as the ServiceProvider which is used in just one place. But we might also be able to use the _oleServiceProvider too at that point since this factory has already been sited?

**Comment by JieCarolHu** [comment]
retest windows_release_unit32_prtest

**Comment by JieCarolHu** [comment]
retest windows_release_unit32_prtest

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview4 ( Once test clears )

**Comment by JieCarolHu** [comment]
retest windows_release_vs-integration_prtest

**Comment by JieCarolHu** [comment]
retest windows_debug_spanish_unit32_prtest

**Comment by sharwell** [comment]
@JieCarolHu Remaining test failures appear to be th eones I'm fixing separately. Shouldn't need to block the merge.
